url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/graph/mxGraphBridge.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * $Id: mxGraphBridge.java,v 1.4 2010-12-01 17:59:41 david Exp $
+ * Copyright (c) 2010, David Benson, Gaudenz Alder
+ */
+package org.jgraph.graph;
+
+import java.awt.geom.Rectangle2D;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.jgraph.JGraph;
+
+import com.jgraph.layout.JGraphFacade;
+import com.mxgraph.model.mxGeometry;
+import com.mxgraph.model.mxGraphModel;
+import com.mxgraph.model.mxIGraphModel;
+import com.mxgraph.util.mxPoint;
+import com.mxgraph.view.mxGraph;
+import com.mxgraph.view.mxGraphView;
+
+/**
+ * Utility class to convert JGraph 5 data structures to JGraph 6 (X). Currently
+ * it is only designed to bridge the JGraph 5 layout facade
+ * 
+ */
+public class mxGraphBridge {
+
+    /**
+     * The mxGraph view
+     */
+    protected mxGraphView mxView;
+
+    /**
+     * The mxGraph
+     */
+    protected mxGraph graph;
+
+    /**
+     * The JGraph UI
+     */
+    protected JGraph jgraph;
+
+    /**
+     * The facade above the JGraph
+     */
+    protected JGraphFacade facade;
+
+    /**
+     * A map to associate mxGraph edges with JGraph edges
+     */
+    protected Map<Object, Object> edges;
+
+    /**
+     * A map to associate mxGraph vertices with JGraph vertices
+     */
+    protected Map<Object, Object> vertices;
+
+    public mxGraphBridge() {
+    }
+
+    public mxGraphBridge(JGraphFacade inputFacade) {
+	this.facade = inputFacade;
+    }
+
+    /**
+     * Populate a fresh mxGraph data model from a JGraph data model
+     */
+    public void syncMx(JGraph inputGraph) {
+	if (inputGraph != null) {
+	    jgraph = inputGraph;
+	} else {
+	    return;
+	}
+
+	GraphModel jModel = jgraph.getModel();
+
+	mxIGraphModel mxModel = null;
+
+	if (graph == null) {
+	    mxModel = new mxGraphModel();
+	    this.graph = new mxGraph(mxModel);
+	}
+
+	mxModel = graph.getModel();
+	mxModel.beginUpdate();
+	try {
+	    int rootCount = jModel.getRootCount();
+
+	    // Insert everything in mx in the same order and
+	    // hierarchy as jgraph
+	    for (int i = 0; i < rootCount; i++) {
+		Object cell = jModel.getRootAt(i);
+
+		if (cell != null) {
+		    insertCell(cell, graph.getDefaultParent());
+		}
+	    }
+
+	    // Edges are inserted without source and targets,
+	    // since the vertices may not exist at insertion
+	    // time
+	    if (edges != null && vertices != null) {
+		Set<Object> keys = edges.keySet();
+
+		for (Object edge : keys) {
+		    Object source = jModel.getSource(edge);
+		    Object target = jModel.getTarget(edge);
+
+		    Object mxEdge = edges.get(edge);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.